test/wallet-rescan-test.js

```diff
@@ -0,0 +1,516 @@
+
+    // Connect MemWallet to chain as minimally as possible
+    node.chain.on('connect', (entry, block) => {
+      alice.addBlock(entry, block.txs);
+    });
+    alice.getNameStatus = async (nameHash) => {
+      assert(Buffer.isBuffer(nameHash));
+      const height = node.chain.height + 1;
+      const state = await node.chain.getNextState();
+      const hardened = state.hasHardening();
+      return node.chain.db.getNameStatus(nameHash, height, hardened);
+    };
+
+    const NAME = rules.grindName(4, 4, network);
+
+    // Hash of the FINALIZE transaction
+    let aliceFinalizeHash;
+
+    async function mineBlocks(n, addr) {
+      addr = addr ? addr : new Address().toString('regtest');
+      const blocks = [];
+      for (let i = 0; i < n; i++) {
+        const block = await node.miner.mineBlock(null, addr);
+        await node.chain.add(block);
+        blocks.push(block);
+      }
+
+      return blocks;
+    }
+
+    before(async () => {
+      await node.open();
+      bob = await wdb.create();
+      bobAddr = await bob.receiveAddress();
+    });
+
+    after(async () => {
+      await node.close();
+    });
+
+    it('should fund wallets', async () => {
+      const blocks = 10;
+      await mineBlocks(blocks, aliceAddr);
+      await mineBlocks(blocks, bobAddr);
+
+      const bobBal  = await bob.getBalance();
+      assert.strictEqual(bobBal.confirmed, blocks * 2000 * 1e6);
+      assert.strictEqual(alice.balance, blocks * 2000 * 1e6);
+    });
+
+    it('should run auction', async () => {
+      // Poor Bob, all he does is send an OPEN but his wallet will
+      // watch all the other activity including TRANSFERS for this name
+      await bob.sendOpen(NAME, true);
+      const openBlocks = await mineBlocks(1);
+      // Coinbase plus open
+      assert.strictEqual(openBlocks[0].txs.length, 2);
+
+      // Advance to bidding phase
+      await mineBlocks(treeInterval);
+      await forValue(alice, 'height', node.chain.height);
+
+      // Alice sends only bid
+      const aliceBid = await alice.createBid(NAME, 20000, 20000);
+      await node.mempool.addTX(aliceBid.toTX());
+      const bidBlocks = await mineBlocks(1);
+      assert.strictEqual(bidBlocks[0].txs.length, 2);
+
+      // Advance to reveal phase
+      await mineBlocks(biddingPeriod);
+      const aliceReveal = await alice.createReveal(NAME);
+      await node.mempool.addTX(aliceReveal.toTX());
+      const revealBlocks = await mineBlocks(1);
+      assert.strictEqual(revealBlocks[0].txs.length, 2);
+
+      // Close auction
+      await mineBlocks(revealPeriod);
+
+      // Alice registers
+      const aliceRegister = await alice.createRegister(
+        NAME,
+        Resource.fromJSON({records:[]})
+      );
+      await node.mempool.addTX(aliceRegister.toTX());
+
+      const registerBlocks = await mineBlocks(1);
+      assert.strictEqual(registerBlocks[0].txs.length, 2);
+    });
+
+    it('should get namestate', async () => {
```

Comment:
it looks like these test blocks are sequentially order dependent, right? this might should be spelled out somewhere up top or at the repo/contributing.md level since it's a general best practice to write test cases that are strictly independent and randomize the order.

Reply:
thats true, we should mention that in contirbuting. In this case I am using separate `describe` blocks for overall test cases, and `it` blocks to verify incrementally. Its a standard we frequently break in this repo.

Reply:
cool, well then yeah just something newcomers should be able to avoid tripping over by viewing the key docs